File: Linked-Lists/206.js
// 206. Reverse Linked List
// Given the head of a singly linked list, reverse the list, and return the reversed list.

var reverseList = function(head) {
    if (head === null || head.next === null) return head
    let walker = head
    let last
    while (walker) {
        walker.prev = walker.next
        if (last) {
            walker.next = last
            last = last.prev
        } else {
            walker.next = null
            last = walker
        }
        if (walker.prev !== null) {
            walker = walker.prev
        } else {
            return walker
        }
    }
};

class ListNode {
    constructor(val, next) {
        this.val = (val===undefined ? 0 : val)
        this.next = (next===undefined ? null : next)
    }
    append(val) {
        const newNode = new ListNode(val)
        if (this.val === null) {
            this.val = newNode
            return
        }
        let walker = this
        while (walker.next) {
            walker = walker.next
        }
        walker.next = newNode
    }
    printAsArray() {
        let array = []
        let walker = this
        while (walker) {
            array.push(walker.val + ' ' + '->')
            walker = walker.next
        }
        array.push('null')
        console.log(array)
    }
}

const testCase = new ListNode(1)
testCase.append(2)
testCase.append(3)
testCase.append(4)
testCase.append(5)
testCase.printAsArray()
reverseList(testCase).printAsArray()

File: Linked-Lists/206.py
# 206. Reverse Linked List
# Given the head of a singly linked list, reverse the list, and return the reversed list.

from typing import Optional

class ListNode:
    def __init__(self, val = 0, next = None):
        self.val = val
        self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head
        prev = None
        while head.next:
            nxt = head.next
            head.next = prev
            prev = head
            head = nxt
        head.next = prev
        return head

node5 = ListNode(5)
node4 = ListNode(4, node5)
node3 = ListNode(3, node4)
node2 = ListNode(2, node3)
test_case = ListNode(1, node2)

s = Solution()
output = s.reverseList(test_case)

# Print list as a string
display = ""
walker = output
while walker:
    display += str(walker.val) + " => "
    walker = walker.next
print(display + "None")